fix(core): Avoid refreshing a host view twice when using transplanted…

atscott opened the pull request:
… views

This change fixes and issue where the expectation was that change detection always goes through `detectChangesInView`. In reality, `detectChangesInternal` directly calls `refreshView` and refreshes a view directly without checking if it was dirty (to my discontent).

This update clears the refresh flags at the top of `refreshView` as well as before traversing child views in `Targeted` mode. The issue only partially exists with signals in that the flags aren't cleared but we don't end up refreshing the view a second time because the `consumerPollProducersForChange` returns `false` the second time we enter `detectChangesInView`.

alxhub reviewed (commented):
How about instead, everywhere we call `refreshView`, we switch to using `detectChangesInternal` but set the `Dirty` flag beforehand?

atscott committed:
fix(core): Avoid refreshing a host view twice when using transplanted views

This change fixes and issue where the expectation was that change
detection always goes through `detectChangesInView`. In reality,
`detectChangesInternal` directly calls `refreshView`
and refreshes a view directly without checking if it was dirty (to my discontent).

This update changes the implementation of `detectChangesInternal` to
actually be "detect changes" not "force refresh of root view and detect
changes". In addition, it adds the refresh flag to APIs that were
previously calling `detectChangesInternal` so we get the same behavior
as before (host view is forced to refresh).

Note that the use of `RefreshView` instead of `Dirty` is _intentional_
here. The `RefreshView` flag is cleared before refreshing the view while
the `Dirty` flag is cleared at the very end. Using the `Dirty` flag
could have consequences because it is a more long-lasting change to the
view flags. Because `detectChangesInView` will immediately clear the
`RefreshView` flag, this change is much more limited and does not
result in a different set of flags during the view refresh.

alxhub commented on packages/core/src/render3/util/change_detection_utils.ts:
```diff
@@ -38,5 +38,6 @@ export function applyChanges(component: {}): void {
  */
 function detectChanges(component: {}): void {
   const view = getComponentViewByInstance(component);
+  view[FLAGS] |= LViewFlags.RefreshView;
```
Given the commit message talks about the importance of `RefreshView` as opposed to `Dirty` here, can you document that in a comment?

alxhub commented:
This PR was merged into the repository by commit 25651218512becb3a491b1c94d643a5066a6d3cb.
